File: testbed_php/test/associations/ManyToMNTest.php
<?php

class ManyToMNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentO(99);
    $this->assertEqual(0,$s->numberOfMentors());
  }
  
  public function test_ConstructorWithTooFew()
  {
  	$this->expectException(new Exception("Unable to create MentorO, must have 2 to 4 students"));
    $s = new StudentO(99);
    new MentorO("blah",array($s));
  }

  public function test_ConstructorWithTooMany()
  {
  	$this->expectException(new Exception("Unable to create MentorO, must have 2 to 4 students"));
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);
    $s5 = new StudentO(95);
    new MentorO("blah",array($s,$s2,$s3,$s4,$s5));
  }
  
  

  public function test_constructorJustBigEnough()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);

    $m = new MentorO("blah",array($s,$s2));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
  }
  

  public function test_constructorJustSmallEnough()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);

    $m = new MentorO("blah",array($s,$s2,$s3,$s4));

    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));
  }  


  public function test_constructorWatchOutForDuplicateEntries()
  {
  	$this->expectException(new Exception("Unable to create MentorO, must have 2 to 4 students"));
    $s = new StudentO(99);
    new MentorO("blah",array($s, $s));
  }
  

  public function test_constructorExistingMentorsOkay()
  {
    $s1 = new StudentO(99);
    $s2 = new StudentO(98);
    $s3 = new StudentO(97);
    
    $m = new MentorO("blah",array($s1,$s2));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m, $s1->getMentor(0));
    $this->assertEqual($m, $s2->getMentor(0));
    $this->assertEqual(0, $s3->numberOfMentors());    

    $m2 = new MentorO("blah2",array($s1,$s2,$s3));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m, $s1->getMentor(0));
    $this->assertEqual($m, $s2->getMentor(0));
    
    $this->assertEqual(3,$m2->numberOfStudents());
    $this->assertEqual($m2, $s1->getMentor(1));
    $this->assertEqual($m2, $s2->getMentor(1));
    $this->assertEqual($m2, $s3->getMentor(0));

  }   
  

  public function test_setStudents_outsideNM()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);
    $s5 = new StudentO(95);

    $m = new MentorO("blah",array($s, $s2));
    
    $this->assertEqual(false,$m->setStudents(array($s2, $s3, $s4, $s5, $s)));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());

    $this->assertEqual(false,$m->setStudents($s5));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
    
    
  }


  public function test_setStudents_doNotAllowDuplicatesNM()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);

    $m = new MentorO("blah",array($s, $s2));
    $this->assertEqual(false,$m->setStudents(array($s2, $s2, $s3)));
  }


  public function test_setStudents_withinNM()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);

    $m = new MentorO("blah",array($s, $s2));
    
    $this->assertEqual(true,$m->setStudents(array($s2, $s3, $s4)));
    $this->assertEqual(3,$m->numberOfStudents());

    $this->assertEqual(0,$s->numberOfMentors());

    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));    
  }
  

  public function test_addMentorTooMany()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);
    $s5 = new StudentO(95);

    $m = new MentorO("blah",array($s,$s2,$s3,$s4));
    
    $this->assertEqual(false,$s5->addMentor($m));
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(0,$s5->numberOfMentors());
    
  }


  public function test_addMentorOkay()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s5 = new StudentO(96);

    $m = new MentorO("blah",array($s,$s2,$s3));
    
    $this->assertEqual(true,$s5->addMentor($m));
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(1,$s5->numberOfMentors());
    
  }
  
  

  public function test_addStudent()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);
    $s5 = new StudentO(95);

    $m = new MentorO("blah",array($s,$s2,$s3));
    
    $this->assertEqual(true,$m->addStudent($s4));
    
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));    
   
    $this->assertEqual(false,$m->addStudent($s5));
    $this->assertEqual(4,$m->numberOfStudents());
  
  }  


  public function test_addStudentHasExistingMentor()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);
    $s5 = new StudentO(95);
    $s6 = new StudentO(94);

    $m = new MentorO("blah",array($s,$s2,$s3));
    $m2 = new MentorO("blah2",array($s4,$s5,$s6));

    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    
    $this->assertEqual($m2,$s4->getMentor(0));
    $this->assertEqual($m2,$s5->getMentor(0));
    $this->assertEqual($m2,$s6->getMentor(0));
    
    $this->assertEqual(true,$m->addStudent($s4));
    
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(3,$m2->numberOfStudents());
    
    $this->assertEqual(2,$s4->numberOfMentors());    
    $this->assertEqual($m2,$s4->getMentor(0));    
    $this->assertEqual($m,$s4->getMentor(1));

    $this->assertEqual(false,$m->addStudent($s5));
  }  


  public function test_removeStudent()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);    
    $s3 = new StudentO(97);
    $s4 = new StudentO(96);

    $m = new MentorO("blah",array($s,$s2,$s3));
    
    $this->assertEqual(false,$m->removeStudent($s4));
    $this->assertEqual(true,$m->removeStudent($s3));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());    

    $this->assertEqual(false,$m->removeStudent($s2));
    $this->assertEqual(2,$m->numberOfStudents());
  }  
  
  
  
  

  public function test_setStudents_empty()
  {
    $s = new StudentO(99);
    $s2 = new StudentO(98);
    $s3 = new StudentO(97);
    $m = new MentorO("blah",array($s,$s2,$s3));

    $this->assertEqual(false,$m->setStudents(array()));
    $this->assertEqual(3,$m->numberOfStudents());
  }
  

  public function test_deletStudentDeletesTheMentorIfNotEnoughLeft()
  {
    $s1 = new StudentO(99);
    $s2 = new StudentO(98);

    $m = new MentorO("blah",array($s1,$s2));
    $studentP = new ProgramO();
    $mentorP = new ProgramO();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(null,$m->getProgram());
    $this->assertEqual(null,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(0,$s1->numberOfMentors());
  }  
  

  public function test_deleteStudentLeavesMentorAloneIfHasEnough()
  {
    $s1 = new StudentO(99);
    $s2 = new StudentO(98);
    $s3 = new StudentO(97);

    $m = new MentorO("blah",array($s1,$s2,$s3));
    $studentP = new ProgramO();
    $mentorP = new ProgramO();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($mentorP,$m->getProgram());
    $this->assertEqual($m,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(0,$s1->numberOfMentors());
  } 
  
}
?>

File: testbed_ruby/test/associations/many_to_m_n_test.rb
require 'test/unit'
require './src-gen-umple/student_o'
require './src-gen-umple/mentor_o'
require './src-gen-umple/program_o'

module CruiseAssociations
class ManyToMNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentO.new(99)
    assert_equal(0,s.number_of_mentors)
  end
  
  def test_ConstructorWithTooFew
    assert_raise RuntimeError do
      s = StudentO.new(99)
      MentorO.new("blah",[s])
    end
  end
  
  def test_ConstructorWithTooMany
    assert_raise RuntimeError do #"Unable to create MentorO, must have 2 to 4 students"
      s = StudentO.new(99)
      s2 = StudentO.new(98)    
      s3 = StudentO.new(97)
      s4 = StudentO.new(96)
      s5 = StudentO.new(95)
      MentorO.new("blah",[s,s2,s3,s4,s5])
    end
  end
  
  def test_constructorJustBigEnough
    s = StudentO.new(99)
    s2 = StudentO.new(98)
  
    m = MentorO.new("blah",[s,s2])
  
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
  end
  
  def test_constructorJustSmallEnough
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
  
    m = MentorO.new("blah",[s,s2,s3,s4])
  
    assert_equal(4,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))
  end  
  
  def test_constructorWatchOutForDuplicateEntries
    assert_raise RuntimeError do #"Unable to create MentorO, must have 2 to 4 students
      s = StudentO.new(99)
      MentorO.new("blah",[s, s])
    end
  end
  
  def test_constructorExistingMentorsOkay
    s1 = StudentO.new(99)
    s2 = StudentO.new(98)
    s3 = StudentO.new(97)
    
    m = MentorO.new("blah",[s1,s2])
  
    assert_equal(2,m.number_of_students)
    assert_equal(m, s1.get_mentor(0))
    assert_equal(m, s2.get_mentor(0))
    assert_equal(0, s3.number_of_mentors)    
  
    m2 = MentorO.new("blah2",[s1,s2,s3])
    
    assert_equal(2,m.number_of_students)
    assert_equal(m, s1.get_mentor(0))
    assert_equal(m, s2.get_mentor(0))
    
    assert_equal(3,m2.number_of_students)
    assert_equal(m2, s1.get_mentor(1))
    assert_equal(m2, s2.get_mentor(1))
    assert_equal(m2, s3.get_mentor(0))
  
  end   
  
  def test_set_students_outsideNM
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
    s5 = StudentO.new(95)
  
    m = MentorO.new("blah",[s, s2])
    
    assert_equal(false,m.set_students([s2, s3, s4, s5, s]))
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
  
    assert_equal(false,m.set_students([s5]))
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
  end
  
  def test_set_students_doNotAllowDuplicatesNM
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
  
    m = MentorO.new("blah",[s, s2])
    assert_equal(false,m.set_students([s2, s2, s3]))
  end
  
  def test_set_students_withinNM
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
  
    m = MentorO.new("blah",[s, s2])
    
    assert_equal(true,m.set_students([s2, s3, s4]))
    assert_equal(3,m.number_of_students)
  
    assert_equal(0,s.number_of_mentors)
  
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))    
  end
  
  def test_add_mentorTooMany
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
    s5 = StudentO.new(95)
  
    m = MentorO.new("blah",[s,s2,s3,s4])
    
    assert_equal(false,s5.add_mentor(m))
    assert_equal(4,m.number_of_students)
    assert_equal(0,s5.number_of_mentors)
    
  end
  
  def test_add_mentorOkay
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s5 = StudentO.new(96)
  
    m = MentorO.new("blah",[s,s2,s3])
    
    assert_equal(true,s5.add_mentor(m))
    assert_equal(4,m.number_of_students)
    assert_equal(1,s5.number_of_mentors)
    
  end

  def test_add_student
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
    s5 = StudentO.new(95)
  
    m = MentorO.new("blah",[s,s2,s3])
    
    assert_equal(true,m.add_student(s4))
    
    assert_equal(4,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))    
   
    assert_equal(false,m.add_student([s5]))
    assert_equal(4,m.number_of_students)
  
  end  
  
  def test_add_studentHasExistingMentor
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
    s5 = StudentO.new(95)
    s6 = StudentO.new(94)
  
    m = MentorO.new("blah",[s,s2,s3])
    m2 = MentorO.new("blah2",[s4,s5,s6])
  
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    
    assert_equal(m2,s4.get_mentor(0))
    assert_equal(m2,s5.get_mentor(0))
    assert_equal(m2,s6.get_mentor(0))
    
    assert_equal(true,m.add_student(s4))
    
    assert_equal(4,m.number_of_students)
    assert_equal(3,m2.number_of_students)
    
    assert_equal(2,s4.number_of_mentors)    
    assert_equal(m2,s4.get_mentor(0))    
    assert_equal(m,s4.get_mentor(1))
  
    assert_equal(false,m.add_student(s5))
  end  
  
  def test_remove_student
    s = StudentO.new(99)
    s2 = StudentO.new(98)    
    s3 = StudentO.new(97)
    s4 = StudentO.new(96)
  
    m = MentorO.new("blah",[s,s2,s3])
    
    assert_equal(false,m.remove_student(s4))
    assert_equal(true,m.remove_student(s3))
    
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)    
  
    assert_equal(false,m.remove_student(s2))
    assert_equal(2,m.number_of_students)
  end  
  
  def test_set_students_empty
    s = StudentO.new(99)
    s2 = StudentO.new(98)
    s3 = StudentO.new(97)
    m = MentorO.new("blah",[s,s2,s3])
  
    assert_equal(false,m.set_students([]))
    assert_equal(3,m.number_of_students)
  end
  
  def test_deletStudentDeletesTheMentorIfNotEnoughLeft
    s1 = StudentO.new(99)
    s2 = StudentO.new(98)
  
    m = MentorO.new("blah",[s1,s2])
    studentP = ProgramO.new
    mentorP = ProgramO.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)
  
    s1.delete
    
    assert_equal(0,m.number_of_students)
    assert_equal(nil,m.get_program)
    assert_equal(nil,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(0,s1.number_of_mentors)
  end  
  
  def test_deleteStudentLeavesMentorAloneIfHasEnough
    s1 = StudentO.new(99)
    s2 = StudentO.new(98)
    s3 = StudentO.new(97)
  
    m = MentorO.new("blah",[s1,s2,s3])
    studentP = ProgramO.new
    mentorP = ProgramO.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)
  
    s1.delete
    
    assert_equal(2,m.number_of_students)
    assert_equal(mentorP,m.get_program)
    assert_equal(m,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(0,s1.number_of_mentors)
  end 
  
end
end